### PR TITLE
Possibly fix an issue with black_death

### PR DESCRIPTION
**Pull request by RedTachyon**
I'm putting this here to possibly get some eyes on this in the meantime. Maybe I'm missing something obvious, but maybe I'm not. The tl;dr is that I think `black_death_v2` might be fundamentally broken, and this is a proposed fix.

The only rough edge I can see around my analysis is what exactly `env.dones` is supposed to be -- particularly if it should contain all agents, or only agents that have been active in the last step or something like that. Which I guess is part of the API description? But idk if it's actually standarized anywhere.

If the API is that `env.dones` is supposed to have a key for every agent that has ever existed (or I guess will ever exist; [this seems to be the case in KAZ at least](https://github.com/Farama-Foundation/PettingZoo/blob/master/pettingzoo/butterfly/knights_archers_zombies/knights_archers_zombies.py#L371)), then I'm like 99% sure the rest of this message is correct, and that the wrapper is broken. The 1% is that it's 2AM at the time of writing.

Obviously this will call for a version bump, so don't merge it yet.

-----
(pasting a message from Discord with a larger explanation because I'm lazy)

I'm running the following code:
```python
obs = env.reset()
for i in trange(1000):
    obs, reward, done, info = env.step({agent_id: env.action_space(agent_id).sample() for agent_id in obs})

    if any(done.values()):
        print(i)
        obs = env.reset()
```

Basically just run a bunch of steps and reset whenever needed. The env is created with this, based on the code from the repo:
```python
def create_env():
    env = knights_archers_zombies_v8.parallel_env()

    env = ss.color_reduction_v0(env, mode="R")
    env = ss.resize_v0(env, x_size=84, y_size=84)
    env = ss.pad_action_space_v0(env)
    env = ss.frame_stack_v1(env, 3)
    env = ss.black_death_v2(env)
    return env
```

And this crashes once every few episodes.

The reason seems to be that one of the agents is actually dead, so the environment tries to get its observation - but there's no observation, so crash.

This should be handled by the black_death wrapper, and I tracked down a very suspicious line in its code (on master)
```python
def observe(self, agent):
    return np.zeros_like(self.observation_space(agent).low) if agent not in self.env.dones else self.env.observe(agent)
```

The idea of this is that if an agent is `done`, then just do zeros; otherwise, get the actual observation.

The problem is that `self.env.dones` is a dictionary that maps each agent to a boolean... and `thing in dict` checks if `thing` is a key in the dictionary. Doesn't matter what its value is. So if we map each agent to whether or not it's alive, this check will just ignore it and pretend all agents are alive. So then it will get the observation for the dead agent. And it will fail. 

It's one of the things where I'm hesitant to believe myself because I think it breaks the wrapper, like, completely? Unless different envs process the dones differently I guess.

### CODE REVIEW

**Comment by benblack769** [comment]
So generally speaking, envs are supposed to remove elements from the dictionary the turn after an agent is done.

```
self.dones[agent] # true
...
self.dones[agent] # false
self.dones[agent] # KeyError
```

The reason the black death waits until the dones are removed is because it is only at that point that the observation is undefined---when the done is still in the dictionary, you should still be able to call `observe(agent)`

The API tests should check for this, but only if the number of steps that the API test takes is sufficiently high to trigger the end of the environment, which might not be the case.

I suspect some of the recent changes to KAZ broke this assumed behavior, and the API tests didn't run for enough cycles to catch the issue.

**Comment by RedTachyon** [comment]
So here's what else I found:
- The [parallel_to_aec_wrapper](https://github.com/Farama-Foundation/PettingZoo/blob/master/pettingzoo/utils/conversions.py#L194) puts only active agents in `dones`
  - As a consequence, envs that go through this conversion either because they start out as parallel, or they go through it due to a wrapper for some reason, will only have active agents in `dones`. This is the case for `KAZ`, and there doesn't seem to be a test for a black_death'd KAZ in the API test suite
- Environments in butterfly that don't feature agents dying like `butterfly.cooperative_pong` (everything but KAZ?) are sloppy in differentiatin between `self.agents` and `self.possible_agents`. Which doessn't necessarily change the behavior, but is impactful in situations like right now.
  - Vast majority of envs don't feature agents dying, so it's hard to find a concrete "typical" approach
- Magent envs are defined as Parallel first, so I'm assuming they keep the `dones` only for active agents because of the wrapper
- MultiWalker doesn't seem to even remove dead agents from `self.agents` from what I can tell? So it also doesn't remove them from `self.dones`. Individual agents dying is an optional feature here and there seems to be something more messed up with it -- when one agent dies, it seems to instantly respawn, at least in terms of giving observations etc. So the same error does not occur, I never managed to get it to a state where `self.agents != self.possible_agents`, even with the right parameters in which a single agent can be `done`
- KAZ (the star here) doesn't remove dead agents from `self.dones`. In fact, it very explicitly uses `self.possible_agents`. This seems to have been added in Farama-Foundation/PettingZoo#602, so pinging @jjshoots (to be clear, I'm 100% not blaming you for anything here, I think I like your approach more, and the whole thing's a mess)
- The [example custom env implementation](https://www.pettingzoo.ml/environment_creation#example-custom-environment) bases its `self.dones` off of `self.agents`, but it never actually updates `self.agents`, so I'm pretty sure it also wouldn't remove dead agents from `self.dones`

tl;dr the feature of agents dying feels barely supported and documented, and there are no environments that explicitly remove an agent from`self.dones`, it only ever happens through the wrapper. There are only two envs where this behavior needs to be specified - MultiWalker which seems to be bugged, and KAZ where it explicitly uses all possible agents. 

This probably calls for a decision, ideally backed by some documentation. If `self.dones` is meant to only contain active agents, then the bug is in KAZ and the wrapper is fine. If `self.dones` is meant to contain possible agents, then the black death wrapper as well as the PZ conversion wrapper are bugged.

For me it's a bit more intuitive that it should contain all agents because we're taking a bit of a high-level view here, but it might be my bias because I first looked at it in the context of KAZ, and frankly, not much is very intuitive about the API.

**Comment by benblack769** [comment]
@RedTachyon  One of the reasons it can't contain *all* of the agents is that it has to support continual environments where agents spawn and die all the time---we are trying to work with Joseph Schwarez to add support to neural MMO.

Belive me, there is a lot of unituitive behavior in the API to me and @jkterry1 as well. However, I suspect when you go and change it to make one use case more intuitive (KAZ in this case), we have found that another use case will be counterintutive.

That is why I recomend all environments that can to implement the parallel environment API, which is easier to understand, and rely on the wrapper to convert it to an AEC API if needed.

It  might be worth going back and changing KAZ and waterworld to implement the parallel environment instead of the AEC environment API.

**Comment by benblack769** [comment]
Another thing I have considered, so that fewer people have to stare at this truely awful black death code, is to write a wrapper in the parallel API, similar to frame-skip: 

https://github.com/Farama-Foundation/SuperSuit/blob/master/supersuit/generic_wrappers/frame_skip.py#L110

**Comment by benblack769** [comment]
Might be worth noting that I originally wrote the parallel_to_aec wrapper concurrently with the magent support specifically so I wouldn't have to think about all the weirdness going on with removing agents.
